allennlp/modules/augmented_lstm.py

```diff
@@ -0,0 +1,171 @@
+
+        inputs, batch_lengths = pad_packed_sequence(inputs, batch_first=True)
+        batch_size = inputs.size()[0]
+        total_timesteps = inputs.size()[1]
+        output_accumulator = Variable(inputs.data.new().resize_(batch_size,
+                                                                total_timesteps,
+                                                                self.output_size).fill_(0))
+        if initial_state is None:
+            full_batch_previous_memory = Variable(torch.zeros([batch_size, self.output_size]))
+            full_batch_previous_state = Variable(torch.zeros([batch_size, self.output_size]))
+        else:
+            full_batch_previous_state = initial_state[0]
+            full_batch_previous_memory = initial_state[1]
+
+        current_length_index = batch_size - 1 if self.direction == "forward" else 0
+        if self.recurrent_dropout_probability > 0.0:
+            dropout_mask = get_dropout_mask(self.recurrent_dropout_probability, [batch_size, self.output_size])
+        else:
+            dropout_mask = None
+
+        for timestep in range(total_timesteps):
+            # The index depends on which end we start.
+            index = timestep if self.direction == "forward" else total_timesteps - timestep - 1
+
+            # What we are doing here is finding the index into the batch dimension
+            # which we need to use for this timestep, because the sequences have
+            # variable length, so once the index is greater than the length of this
+            # particular batch sequence, we no longer need to do the computation for
+            # this sequence. The key thing to recognise here is that _the batch inputs
+            # must be _ordered_ by length from longest (first in batch) to shortest
+            # (last) so initially, we are going forwards with every sequence and as we
+            # pass the index at which the shortest elements of the batch finish,
+            # we stop picking them up for the computation.
+            if self.direction == "forward":
+                while batch_lengths[current_length_index] <= index:
+                    current_length_index -= 1
+            # If we're going backwards, we are _picking up_ more indices.
+            elif self.direction == "backward":
+                # First conditional: Are we already at the maximum number of elements in the batch?
+                # Second conditional: Does the next shortest sequence beyond the current batch
+                # index require computation use this timestep?
+                while current_length_index < (len(batch_lengths) - 1) and \
+                                batch_lengths[current_length_index + 1] > index:
+                    current_length_index += 1
+
+            # Actually get the slices of the batch which we need for the computation at this timestep.
+            previous_memory = full_batch_previous_memory[0: current_length_index + 1].clone()
```

Comment:
The space after the `:` here is confusing to me, here and below.  See http://legacy.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements.